### DISCUSSION TITLE
.NET 4.6 Web App convert HEIC to JPEG

### DISCUSSION THREAD

**Discussion by jwalker5006** [dlemstra/Magick.NET#738] [Help]
I have a web app that has a file upload form with an end goal of saving the uploaded file as a jpeg when the filetype is heic.

Because System.Drawing.Imaging does NOT support heic, and it appears Magick does (possible HEIC requires the libheif delegate library.) not sure how you would do this in a C# or VB.Net web app, or if I actually need the library for converting a uploaded file from HEIC to Jpeg.

What I am doing now is taking the uploaded file, convert to Byte() then put in a MemoryStream, then try and MagickImage it with a settings file where I tell Magick that the in file format is heic. ( had to specify the format to get past the "no decode delegate for this image format `' @ error/blob.c/BlobToImage/458" error). 

Now I get a "Invalid input: No 'ftyp' box `' @ error/heic.c/IsHeifSuccess/134" which searching the Magick github I see no code reference to ftpy, but when I search the [library that magick says is required,](https://imagemagick.org/script/formats.php) [I do see references ](https://github.com/strukturag/libheif/search?q=ftyp)to this. But the libheif seems to be only c++ code, and have no clue how to use that library in a .NET web app.

Here is the code I am using to try and convert the heic byte data to jpeg byte data. Error occurs at the MagickImage line.

**VB**
```
       Dim DataBytes As Byte()
        Dim settings As MagickReadSettings = New MagickReadSettings()
        settings.Format = MagickFormat.Heic


        Using m As MemoryStream = New MemoryStream()
            Using image As MagickImage = New MagickImage(heicfile, settings)
                image.Format = MagickFormat.Jpeg
                image.Write(m)
                DataBytes = image.ToByteArray()
            End Using
            Return Convert.ToBase64String(DataBytes)
        End Using
```

**C#**
```
{
    byte[] DataBytes;
    var settings = new MagickReadSettings();
    settings.Format = MagickFormat.Heic;
    using (var m = new MemoryStream())
    {
        using (var image = new MagickImage(heicfile, settings))
        {
            image.Format = MagickFormat.Jpeg;
            image.Write(m);
            DataBytes = image.ToByteArray();
        }

        return Convert.ToBase64String(DataBytes);
    }
}
```

**Comment by dlemstra** [comment]
This could be an issue with your file or an issue with how you create the byte array. Can you share your input image?

**Reply by jwalker5006** [reply]
The file was fine, you were 100% on the second point. The byte array was not created correctly. Thank you.

Below is the .NET code I used that worked. Its VB, but would work just the same in C#. Upload control ID is FileUpload1. This code worked like a charm. Fast and error free.

## In the event handler for the save image button, the input stream is converted into a byte array, then I pass the heic byte array into a function that resizes and converts to the heic to a jpeg
```
                    Dim strmStream As Stream = FileUpload1.PostedFile.InputStream
                    Dim intFileLength As Int32 = strmStream.Length
                    Dim bytUpfile As Byte() = New Byte(intFileLength - 1) {}
                    strmStream.Read(bytUpfile, 0, intFileLength)
                    strmStream.Close()

                   bytUpfile = ConvertHeicToJpg(bytUpfile)
```

## The resize and conversion function
```
    Function ConvertHeicToJpg(heicfile As Byte()) As Byte()
        Dim DataBytes As Byte()
        Dim settings As MagickReadSettings = New MagickReadSettings()
        settings.Format = MagickFormat.Heic

        Using m As MemoryStream = New MemoryStream()
            Using image As MagickImage = New MagickImage(heicfile, settings)
                image.Format = MagickFormat.Jpeg
                Dim size = New MagickGeometry(1024, 0)
                image.Resize(size)
                image.Write(m)
                DataBytes = image.ToByteArray()
            End Using
        End Using

        Return DataBytes
    End Function
```

I believe I could have also skipped the MagickGeometry(1024,0) and just done image.Resize(1024,0)
